File: DSA/VivaswanSingh_590023022/Day_34/q2.c
#include <stdio.h>

int maxVisiblePeople(int arr[], int n)
{
    int left[n], right[n];
    int stack[n];
    int top = -1;

    // left
    for (int i = 0; i < n; i++)
    {
        while (top >= 0 && arr[stack[top]] < arr[i])
            top--;

        if (top == -1)
            left[i] = i;
        else
            left[i] = i - stack[top];

        stack[++top] = i;
    }

    
    top = -1;

    // right
    for (int i = n - 1; i >= 0; i--)
    {
        while (top >= 0 && arr[stack[top]] < arr[i])
            top--;

        if (top == -1)
            right[i] = n - 1 - i;
        else
            right[i] = stack[top] - i;

        stack[++top] = i;
    }

    int ans = 1;

    for (int i = 0; i < n; i++)
    {
        int visible = left[i] + right[i] + 1;
        if (visible > ans)
            ans = visible;
    }

    return ans;
}

int main()
{
    int arr[] = {6, 2, 5, 4, 5, 1, 6};
    int n = sizeof(arr) / sizeof(arr[0]);

    printf("%d\n", maxVisiblePeople(arr, n));

    return 0;
}

File: DSA/vaishnavi_joshi_590016947/Day34/Q2.java
import java.util.*;

public class Main {

    public static int maximumPeopleVisible(int[] arr) {
        int n = arr.length;

        int[] leftGreater = new int[n];
        int[] rightGreater = new int[n];

        Stack<Integer> stack = new Stack<>();

        // Nearest Greater or Equal on Left
        for (int i = 0; i < n; i++) {
            while (!stack.isEmpty() && arr[stack.peek()] < arr[i]) {
                stack.pop();
            }

            if (stack.isEmpty())
                leftGreater[i] = -1;
            else
                leftGreater[i] = stack.peek();

            stack.push(i);
        }

        stack.clear();

        // Nearest Greater or Equal on Right
        for (int i = n - 1; i >= 0; i--) {
            while (!stack.isEmpty() && arr[stack.peek()] < arr[i]) {
                stack.pop();
            }

            if (stack.isEmpty())
                rightGreater[i] = n;
            else
                rightGreater[i] = stack.peek();

            stack.push(i);
        }

        int ans = 0;

        for (int i = 0; i < n; i++) {
            int visible = (i - leftGreater[i] - 1)
                        + (rightGreater[i] - i - 1)
                        + 1;

            ans = Math.max(ans, visible);
        }

        return ans;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        int n = sc.nextInt();
        int[] arr = new int[n];

        for (int i = 0; i < n; i++)
            arr[i] = sc.nextInt();

        System.out.println(maximumPeopleVisible(arr));
    }
}